File: testing/ecosystem/python/neo4j_bolt_smoke.py
#!/usr/bin/env python3
import json
import os

from neo4j import GraphDatabase


def main() -> int:
    uri = os.environ["NEO4J_URI"]
    user = os.environ["NEO4J_USER"]
    password = os.environ["NEO4J_PASSWORD"]

    driver = GraphDatabase.driver(uri, auth=(user, password), connection_timeout=5.0)
    try:
        with driver.session() as session:
            result = session.run("RETURN 1 AS one, 'ok' AS status")
            if result.keys() != ["one", "status"]:
                raise RuntimeError(f"unexpected result keys: {result.keys()!r}")
            record = result.single()
            if record is None:
                raise RuntimeError("neo4j driver returned no row for RETURN probe")
            one = record["one"]
            status = record["status"]
            if one != 1 or status != "ok":
                raise RuntimeError(
                    f"unexpected neo4j driver payload: one={one!r}, status={status!r}"
                )

            param_record = session.run(
                "RETURN $one AS one, $status AS status",
                one=1,
                status="ok",
            ).single()
            if param_record is None or param_record["one"] != 1 or param_record["status"] != "ok":
                raise RuntimeError(f"unexpected parameter payload: {param_record!r}")

            unwind_record = session.run(
                "UNWIND [1, 2, 3] AS x RETURN sum(x) AS total"
            ).single()
            if unwind_record is None or unwind_record["total"] != 6:
                raise RuntimeError(f"unexpected UNWIND payload: {unwind_record!r}")

            try:
                session.run("RETURN )").consume()
            except Exception as exc:
                error_text = str(exc)
                if not error_text:
                    raise RuntimeError("cypher syntax error returned an empty exception") from exc
            else:
                raise RuntimeError("expected cypher syntax error was not raised")
    finally:
        driver.close()

    print(
        json.dumps(
            {
                "details": "Neo4j Python driver connected over Bolt and completed read-only RETURN, parameter, UNWIND and error probes",
                "checks": [
                    "bolt_connect",
                    "auth",
                    "session",
                    "return_probe",
                    "result_keys",
                    "parameters",
                    "unwind",
                    "cypher_error",
                ],
                "queries": [
                    "return_probe",
                    "parameter_probe",
                    "unwind_probe",
                    "cypher_error_probe",
                ],
                "capabilities": {
                    "read_only": True,
                    "write_supported": False,
                    "access_mode": "read",
                    "database_scope": "default",
                },
                "uri": uri,
            }
        )
    )
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:  # pragma: no cover - smoke script
        print(json.dumps({"error": str(exc)}))
        raise

File: testing/ecosystem/java/Neo4jBoltSmoke.java
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Record;
import org.neo4j.driver.Result;
import org.neo4j.driver.Session;

public final class Neo4jBoltSmoke {
    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.println("usage: Neo4jBoltSmoke <bolt-uri> <user> <password>");
            System.exit(2);
        }

        String uri = args[0];
        String user = args[1];
        String password = args[2];

        try (Driver driver = GraphDatabase.driver(uri, AuthTokens.basic(user, password));
             Session session = driver.session()) {
            Result result = session.run("RETURN 1 AS one, 'ok' AS status");
            if (!result.hasNext()) {
                throw new IllegalStateException("neo4j-java-driver returned no row for RETURN probe");
            }
            Record record = result.next();
            long one = record.get("one").asLong();
            String status = record.get("status").asString();
            if (one != 1L || !"ok".equals(status)) {
                throw new IllegalStateException(
                    "unexpected neo4j-java-driver payload: one=" + one + ", status=" + status
                );
            }

            System.out.println("{\"details\":\"Neo4j Java driver connected over Bolt and completed a read-only RETURN probe\",\"checks\":[\"bolt_connect\",\"auth\",\"session\",\"return_probe\"]}");
        } catch (Exception error) {
            System.out.println("{\"error\":" + quoteJson(error.getMessage()) + "}");
            System.exit(1);
        }
    }

    private static String quoteJson(String value) {
        if (value == null) {
            return "null";
        }
        StringBuilder out = new StringBuilder(value.length() + 2);
        out.append('"');
        for (int i = 0; i < value.length(); i++) {
            char ch = value.charAt(i);
            switch (ch) {
                case '"':
                    out.append("\\\"");
                    break;
                case '\\':
                    out.append("\\\\");
                    break;
                case '\b':
                    out.append("\\b");
                    break;
                case '\f':
                    out.append("\\f");
                    break;
                case '\n':
                    out.append("\\n");
                    break;
                case '\r':
                    out.append("\\r");
                    break;
                case '\t':
                    out.append("\\t");
                    break;
                default:
                    if (ch < 0x20) {
                        out.append(String.format("\\u%04x", (int) ch));
                    } else {
                        out.append(ch);
                    }
            }
        }
        out.append('"');
        return out.toString();
    }
}
